src/control/drpc/module_svc_test.go

```diff
@@ -0,0 +1,252 @@
+
+// mockModule is a test module used for verifying the drpc.ModuleService
+type mockModule struct {
+	HandleCallResponse []byte
+	HandleCallErr      error
+	IDValue            int32
+}
+
+func (m *mockModule) HandleCall(client *Client, method int32, input []byte) ([]byte, error) {
+	return m.HandleCallResponse, m.HandleCallErr
+}
+
+func (m *mockModule) InitModule(state ModuleState) {}
+
+func (m *mockModule) ID() int32 {
+	return m.IDValue
+}
+
+func newTestModule(ID int32) *mockModule {
+	return &mockModule{
+		IDValue: ID,
+	}
+}
+
+func TestNewModuleService(t *testing.T) {
+	log, buf := logging.NewTestLogger(t.Name())
+	defer common.ShowBufferOnFailure(t, buf)
+
+	service := NewModuleService(log)
+
+	if service == nil {
+		t.Fatal("service was nil")
+	}
+
+	common.AssertEqual(t, len(service.modules), 0, "expected empty module list")
+}
+
+func TestService_RegisterModule_Single_Success(t *testing.T) {
+	log, buf := logging.NewTestLogger(t.Name())
+	defer common.ShowBufferOnFailure(t, buf)
+
+	service := NewModuleService(log)
+	expectedID := defaultTestModID
+	testMod := newTestModule(expectedID)
+
+	err := service.RegisterModule(testMod)
+
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+
+	mod, ok := service.modules[expectedID]
+
+	if !ok {
+		t.Fatalf("module wasn't found under ID %d", expectedID)
+	}
+
+	if diff := cmp.Diff(testMod, mod); diff != "" {
+		t.Fatalf("(-want, +got)\n%s", diff)
+	}
+}
+
+func TestService_RegisterModule_Multiple_Success(t *testing.T) {
+	log, buf := logging.NewTestLogger(t.Name())
+	defer common.ShowBufferOnFailure(t, buf)
+
+	service := NewModuleService(log)
+	expectedIDs := []int32{-1, 2, 255, defaultTestModID}
+	testMods := make([]*mockModule, 0, len(expectedIDs))
+
+	for _, id := range expectedIDs {
+		mod := newTestModule(id)
+		testMods = append(testMods, mod)
+
+		err := service.RegisterModule(mod)
+
+		if err != nil {
+			t.Fatalf("expected no error, got: %v", err)
+		}
+	}
+
+	for i, id := range expectedIDs {
+		mod, ok := service.modules[id]
+
+		if !ok {
+			t.Fatalf("registered module %d wasn't found", id)
+		}
+
+		if diff := cmp.Diff(testMods[i], mod); diff != "" {
+			t.Fatalf("(-want, +got)\n%s", diff)
+		}
+	}
+}
+
+func TestService_RegisterModule_DuplicateID(t *testing.T) {
+	log, buf := logging.NewTestLogger(t.Name())
+	defer common.ShowBufferOnFailure(t, buf)
+
+	service := NewModuleService(log)
+	testMod := newTestModule(15)
+	dupMod := newTestModule(testMod.IDValue)
+
+	err := service.RegisterModule(testMod)
+	if err != nil {
+		t.Fatalf("expected no error for initial registration, got: %v",
+			err)
+	}
+
+	err = service.RegisterModule(dupMod)
+
+	common.CmpErr(t, errors.New("already exists"), err)
+}
+
+func getGarbageBytes() []byte {
+	badBytes := make([]byte, 250)
+	for i := range badBytes {
+		badBytes[i] = byte(i)
+	}
+
+	return badBytes
+}
+
+func getCallBytes(t *testing.T, sequence int64, moduleID int32) []byte {
+	call := &Call{
```

Comment:
It's usually a good practice to add t.Helper() to test functions which call Fatal and friends. This will adjust the call stack so that the failure is attributed to the test that called the helper.

Compare the following failed test outputs:
```
    --- FAIL: TestProviderFormat/dcpm:_mountpoint_doesn't_exist;_not_formatted (0.00s)
        provider_test.go:802: unexpected response (-want, +got):
              &scm.FormatResponse{
                Mountpoint: "/tmp/TestProviderFormat-dcpm:_mountpoint_doesn't_exist;_not_formatted420401907/mnt/daos",
            -   Formatted:  false,
            +   Formatted:  true,
              }
```
vs
```
    --- FAIL: TestProviderFormat/dcpm:_mountpoint_doesn't_exist;_not_formatted (0.00s)
        provider_test.go:758: unexpected response (-want, +got):
              &scm.FormatResponse{
                Mountpoint: "/tmp/TestProviderFormat-dcpm:_mountpoint_doesn't_exist;_not_formatted587226392/mnt/daos",
            -   Formatted:  false,
            +   Formatted:  true,
              }
```
It's a subtle difference, but notice in the second example how the line number is 758, which corresponds to the cmpRes() test helper vs. 802 which is where cmpRes() is called. Personally, I prefer the first one because when I open my editor to look at the code, I can just go to line 802 and see what the test is doing. I can dig into the helper if I need to, but usually I want to start in the failed test body. It's not a hard rule, just something to consider when writing the tests and deciding what is going to be easiest for someone to decipher when things fail later on.

Reply:
Thanks for the info - that is very useful.